src/panel/direction/MobileRoadMapPreview.jsx

```diff
@@ -2,62 +2,80 @@ import React from 'react';
 import PropTypes from 'prop-types';
 import RoadMapStep from './RoadMapStep';
 import { fire } from 'src/libs/customEvents';
-import { CloseButton } from 'src/components/ui';
+import classnames from 'classnames';
+import { Button } from '../../components/ui';
 
 export default class MobileRoadMapPreview extends React.Component {
+  constructor(props) {
+    super(props);
+    this.stepsRef = React.createRef();
+  }
+
   static propTypes = {
     steps: PropTypes.array.isRequired,
     onClose: PropTypes.func.isRequired,
-  }
+  };
 
   state = {
     currentStep: 0,
-  }
+    expanded: false,
+  };
 
   componentDidMount() {
-    this.zoomToCurrentStep();
-    fire('move_mobile_geolocation_button', 65);
-  }
-
-  zoomToCurrentStep = () => {
+    fire('move_mobile_bottom_ui', 105);
     fire('zoom_step', this.props.steps[this.state.currentStep]);
   }
 
-  previous = () => {
-    this.setState(
-      prevState => ({ currentStep: Math.max(prevState.currentStep - 1, 0) }),
-      this.zoomToCurrentStep
-    );
-  }
+  scroll = () => {
 
-  next = () => {
-    this.setState(
-      prevState => ({ currentStep: Math.min(prevState.currentStep + 1, this.props.steps.length) }),
-      this.zoomToCurrentStep
+    const currentStep = Math.floor(
+      // Divide the step container's scrollLeft up to the middle of the screen with the size of a step
+      // to determine which step is present at the middle of the screen
+      (this.stepsRef.current.scrollLeft + window.innerWidth / 2) / (window.innerWidth - 70 + 12)
     );
-  }
 
-  render() {
-    const currentStep = this.props.steps[this.state.currentStep];
+    // If it has changed, save it and highlight it as the current step
+    if (this.state.currentStep !== currentStep) {
+      this.setState({ currentStep });
+      fire('zoom_step', this.props.steps[currentStep]);
+    }
+  };
 
-    return <React.Fragment>
-      <div className="itinerary_mobile_step">
-        <CloseButton onClick={this.props.onClose} />
-        <RoadMapStep step={currentStep} />
-      </div>
+  toggleSize = () => {
+    this.setState( { expanded: !this.state.expanded } );
+    fire('move_mobile_bottom_ui', this.stepsRef.current.offsetHeight);
```

Comment:
`setState` and the triggered `render` can be async, so here you're not guaranteed that the `offsetHeight` will be the new one.
This is what you see in action here, where things are desynchronized.
![Peek 08-12-2020 12-07](https://user-images.githubusercontent.com/243653/101476406-fcb64200-394d-11eb-9b89-7cb67182baad.gif)

To solve this, we must be sure before firing the event that the component has been updated after the `setState`. This is what the `componentDidUpdate` is for, so you can use it to check what state value has changed and trigger the right events while being sure the render pass has been made.

This is one solution, but this is the "old way". I think this is the good occasion to convert this component to a functional one, when such actions (actually side-effects) can be written much more naturally:
```js
const [expanded, setExpanded] = useState(false); // functional way of declaring state
// ...
useEffect(
    () => { /* ... */  }, // do what you want…
    [expanded]  // … everytime "expanded" changes
);  
```

This would require other changes to the component, but I think it's worth it.
Would you want to try?